mono/metadata/threads.c

```diff
@@ -4849,9 +4849,9 @@ self_abort_internal (void)
 
 	exc = mono_thread_request_interruption (TRUE);
 	if (exc)
-		mono_raise_exception (exc);
-
-	mono_thread_info_self_interrupt ();
+		mono_set_pending_exception (exc);
```

Comment:
That doesn't really solve the embedders problem, though; it just pushes the error up one level.

I think in this case we'll either have to keep throwing (not great) or else add a callback (that defaults to throwing).  I could easily imagine API clients doing something like:
```
   mono_thread_stop (self);
   g_assert ("unreachable");
```